compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt

```diff
@@ -29,19 +31,36 @@ object FirSealedSupertypeChecker : FirClassChecker() {
     }
 
     private fun checkGlobalDeclaration(declaration: FirClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        for (it in declaration.superTypeRefs) {
-            val classId = it.coneType.classId ?: continue
+        val subclassPackage = declaration.classId.packageFqName
+        val subClassSession = declaration.moduleData
+        for (superTypeRef in declaration.superTypeRefs) {
+            val classId = superTypeRef.coneType.classId ?: continue
 
             if (classId.isLocal) {
                 continue
             }
 
-            val classSymbol = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
+            val superClass = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
 
-            if (classSymbol.modality == Modality.SEALED && declaration.classId.packageFqName != classSymbol.classId.packageFqName) {
-                reporter.reportOn(it.source, FirErrors.SEALED_SUPERTYPE, context)
+            if (!superClass.isSealed) continue
+            if (superClass.origin == FirDeclarationOrigin.Java) {
+                reporter.reportOn(superTypeRef.source, FirErrors.CLASS_INHERITS_JAVA_SEALED_CLASS, context)
                 continue
             }
+            val superClassPackage = superClass.classId.packageFqName
+            if (superClassPackage != subclassPackage) {
+                reporter.reportOn(
+                    superTypeRef.source,
+                    FirErrors.SEALED_INHERITOR_IN_DIFFERENT_PACKAGE,
+                    subclassPackage,
+                    superClassPackage,
+                    context
+                )
+            }
+            if (superClass.moduleData != subClassSession) {
```

Comment:
Better to use `declaration.moduleData` directly here since it's only used here. Or at least rename `subClassSession` to `subclassModuleData` since it's not the session anymore 🙂

Reply:
🤦